fix: remove old nvidia drivers when doing updates

faiq opened the pull request:
**What problem does this PR solve?**:

This checks to see if the system had previous versions of the nvidia driver installed by dkp. If `nvidia_driver_version` provided by the user is newer it removes and uninstalls the old driver and stops the `nvidia-persistenced` systemd unit before upgrading to the newer driver.


**Which issue(s) does this PR fix?**:
<!-- Add a link to the JIRA issue for both items below
* jql=key in (D2IQ-NUMBER)
-->
* https://jira.d2iq.com/browse/D2IQ-NUMBER


**Special notes for your reviewer**:
<!--
Use this to provide any additional information to the reviewers.
This may include:
- Manual testing steps.
- Best way to review the PR.
- Where the author wants the most review attention on.
- etc.
-->


**Does this PR introduce a user-facing change?**:
<!--
If yes, add a message in the 'release-note' block below.
If this PR fixes a COPS ticket, include it after the note like: "CLI: Some bug fix. (COPS-xxxx)"
-->
```release-note

```

faiq commented on ansible/roles/gpu/tasks/nvidia-gpu.yaml:
```diff
@@ -48,6 +48,58 @@
+- name: extract version from file names
+  set_fact:
+    nvidia_driver_versions: "{{ nvidia_driver_files.files | map(attribute='path') | map('regex_replace', '^.*NVIDIA-Linux-x86_64-(\\d+\\.\\d+\\.\\d+)\\.run$', '\\1') | list }}"
+    uninstall: false
```
lazily sets the fact here to false so if there is no version < nvidia_driver_version the next block doesn't complain about `uninstall` not being set.

faiq commented:
I ran this against a machine that already had 470 drivers installed. I ran it with the following diff in the vars to bump the driver version to 535.171.04

```
$ git diff group_vars/all/defaults.yaml
diff --git a/ansible/group_vars/all/defaults.yaml b/ansible/group_vars/all/defaults.yaml
index 5b73d66..817399a 100644
--- a/ansible/group_vars/all/defaults.yaml
+++ b/ansible/group_vars/all/defaults.yaml
@@ -105,7 +105,7 @@ nvidia_remote_bundle_path: "/opt/dkp/nvidia"
 
 # NOTE(jkoelker) UEK is the default. Set to RHCK to change to RH kernel.
 oracle_kernel: UEK
-nvidia_driver_version: "470.199.02"
+nvidia_driver_version: "535.171.04"
 nvidia_runfile_installer: "NVIDIA-Linux-x86_64-{{ nvidia_driver_version }}.run"
 nvidia_runfile_installer_url: "https://download.nvidia.com/XFree86/Linux-x86_64/{{ nvidia_driver_version }}/{{ nvidia_runfile_installer }}"
 suse_packagehub_product: PackageHub/{{ ansible_distribution_version }}/{{ ansible_architecture }}
```

```bash
    amazon-ebs.kib_image: TASK [gpu : find nvidia driver installer file] *********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : extract version from file names] ***********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : compare versions with nvidia_driver_version] ***********************
    amazon-ebs.kib_image: ok: [default] => (item=470.239.06)
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : uninstall older versions] ******************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : stop persistenced service before installing the new drivers] *******
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : uninstall older versions] ******************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : remove older nvidia driver directory] ******************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : remove older runfile] **********************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : download gpu runfile] **********************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : extract driver source files] ***************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : install driver] ****************************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Create ld.so conf dir] *********************************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Create ld.so conf for nvidia] **************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Write location of library files] ***********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : run nvidia-smi] ****************************************************
    amazon-ebs.kib_image: changed: [default]
```

supershal commented on ansible/roles/gpu/tasks/nvidia-gpu.yaml:
```diff
@@ -48,6 +48,58 @@
+      state: stopped
+    become: yes
+
+  - name: uninstall older versions
```
This task (`uninstall older versions`) is executed again after stopping the service. is that intentional?

faiq replied:
fixed

supershal commented on ansible/roles/gpu/tasks/nvidia-gpu.yaml:
```diff
@@ -48,6 +48,58 @@
+    previous_version: "{{ item }}"
+
+- name: uninstall and stop programs using older nvidia drivers
+  block:
```
Will this block only executed once for a last `previous_version` .  If intention is to remove all older versions, I think this block should be in the loop (not sure its possible)

faiq replied:
there should only be one older version at a time, but i pushed up some code to do it in a loop in case there are others

supershal replied:
Thanks faiq. it looks neat.

faiq commented:
after pushing the code changes i got this 

```
 amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : create nvidia install directory] ***********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : find nvidia driver installer file] *********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : extract version from file names] ***********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : compare versions with nvidia_driver_version] ***********************
    amazon-ebs.kib_image: ok: [default] => (item=470.239.06)
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : uninstall older versions] ******************************************
    amazon-ebs.kib_image: changed: [default] => (item=470.239.06)
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : stop persistenced service before installing the new drivers] *******
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : remove older nvidia driver directory] ******************************
    amazon-ebs.kib_image: changed: [default] => (item=470.239.06)
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : remove older runfile] **********************************************
    amazon-ebs.kib_image: changed: [default] => (item=470.239.06)
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : download gpu runfile] **********************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : extract driver source files] ***************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : install driver] ****************************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Create ld.so conf dir] *********************************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Create ld.so conf for nvidia] **************************************
    amazon-ebs.kib_image: changed: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : Write location of library files] ***********************************
    amazon-ebs.kib_image: ok: [default]
    amazon-ebs.kib_image:
    amazon-ebs.kib_image: TASK [gpu : run nvidia-smi] ****************************************************
    amazon-ebs.kib_image: changed: [default]
```

supershal reviewed (approved):
Looks great.
